pallets/registrar/src/lib.rs

```diff
@@ -562,11 +745,108 @@ pub mod pallet {
 
             <PendingParaIds<T>>::put(future);
 
+            let pending_paused = <PendingPaused<T>>::get();
+
+            let (mut past_and_present, future) =
+                pending_paused
+                    .into_iter()
+                    .partition::<Vec<_>, _>(|&(apply_at_session, _)| {
+                        apply_at_session <= *session_index
+                    });
+
+            if past_and_present.len() > 1 {
+                // This should never happen since we schedule parachain changes only into the future
+                // sessions and this handler called for each session change.
+                log::error!(
+                    target: LOG_TARGET,
+                    "Skipping applying paused parachain changes scheduled sessions in the past",
+                );
+            }
+
+            let new_paused = past_and_present.pop().map(|(_, paras)| paras);
+            if let Some(ref new_paused) = new_paused {
+                // Apply the new parachain list.
+                Paused::<T>::put(new_paused);
+            }
+
+            <PendingPaused<T>>::put(future);
+
+            let pending_paras = <PendingToRemove<T>>::get();
+
+            let (mut past_and_present, future) =
+                pending_paras
+                    .into_iter()
+                    .partition::<Vec<_>, _>(|&(apply_at_session, _)| {
+                        apply_at_session <= *session_index
+                    });
+
+            if past_and_present.len() > 1 {
+                // This should never happen since we schedule parachain changes only into the future
+                // sessions and this handler called for each session change.
+                log::error!(
+                    target: LOG_TARGET,
+                    "Skipping applying to_remove parachain changes scheduled sessions in the past",
+                );
+            }
+
+            let new_paras = past_and_present.pop().map(|(_, paras)| paras);
+            if let Some(ref new_paras) = new_paras {
+                for para_id in new_paras {
+                    Self::cleanup_deregistered_para_id(*para_id);
+                }
+            }
+
+            <PendingToRemove<T>>::put(future);
+
             SessionChangeOutcome {
                 prev_paras,
                 new_paras,
             }
         }
+
+        /// Remove all para id storage in this pallet,
+        /// and execute para_deregistered hook to clean up other pallets as well
+        fn cleanup_deregistered_para_id(para_id: ParaId) {
+            ParaGenesisData::<T>::remove(para_id);
+            BootNodes::<T>::remove(para_id);
+            T::RegistrarHooks::para_deregistered(para_id);
+        }
+
+        fn schedule_parachain_cleanup(para_id: ParaId) -> DispatchResult {
+            let mut pending_paras = PendingToRemove::<T>::get();
+            // First, we need to decide what we should use as the base paras.
+            let mut base_paras = pending_paras
```

Comment:
This is wrong, add a test where we deregister 2 container chains, 1 in session 2 and 1 in session 3:

```
State after deregister(2000):
(2, [2000])

Wait 1 session
State after deregister(2001):
(2, [2000])
(3, [2000, 2001])
```

Container chain 2000 will be cleaned up twice. Also add a test where they are deregistered in the same session.

Reply:
Done, in `deregister_2_container_chains_in_consecutive_sessions` and `deregister_2_container_chains_in_same_block`